src/stan/io/deserializer.hpp

```diff
@@ -0,0 +1,1978 @@
+   * @param sizes A parameter pack of integral types representing the
+   *  dimensions of the `var_value` matrix or vector.
+   */
+  template <typename Ret, typename T_ = T, typename... Sizes,
+            require_var_t<T_>* = nullptr, require_var_matrix_t<Ret>* = nullptr>
+  auto read(Sizes... sizes) {
+    using value_t = typename std::decay_t<Ret>::value_type;
+    using var_v_t = stan::math::promote_scalar_t<stan::math::var, value_t>;
+    return stan::math::to_var_value(this->read<var_v_t>(sizes...));
+  }
+
+  /**
+   * Return an Eigen type when the deserializers inner class is not var.
+   * @tparam Ret The type to return.
+   * @tparam T_ Should never be set by user, set to default value of `T` for
+   *  performing deduction on the class's inner type.
+   * @tparam Sizes A parameter pack of integral types.
+   * @param sizes A parameter pack of integral types representing the
+   *  dimensions of the `var_value` matrix or vector.
+   */
+  template <typename Ret, typename T_ = T, typename... Sizes,
+            require_not_var_t<T_>* = nullptr,
+            require_var_matrix_t<Ret>* = nullptr>
+  auto read(Sizes... sizes) {
+    using value_t = typename std::decay_t<Ret>::value_type;
+    return this->read<value_t>(sizes...);
+  }
+
+  /**
+   * Return an `std::vector`
+   * @tparam Ret The type to return.
+   * @tparam Size an integral type.
+   * @tparam Sizes integral types.
+   * @param m The size of the vector.
+   * @param dims a possible set of inner container sizes passed to subsequent
+   * `read` functions.
+   */
+  template <typename Ret, typename Size, typename... Sizes,
+            require_std_vector_t<Ret>* = nullptr>
+  inline auto read(Size m, Sizes... dims) {
+    if (unlikely(m == 0)) {
+      return Ret();
+    } else {
+      std::decay_t<Ret> ret_vec;
+      ret_vec.reserve(m);
+      for (Size i = 0; i < m; ++i) {
+        ret_vec.emplace_back(this->read<value_type_t<Ret>>(dims...));
+      }
+      return ret_vec;
+    }
+  }
+
+  /**
+   * Return the next object, checking that it's elements are
+   * greater than or equal to the specified lower bound.
+   *
+   * <p>See <code>stan::math::check_greater_or_equal(T,double)</code>.
+   *
+   * @tparam Ret The type to return.
+   * @tparam LB Type of lower bound.
+   * @tparam Sizes A pack of possible sizes to construct the object from.
+   * @param lb Lower bound.
+   * @param sizes a pack of sizes to use to construct the return.
+   * @throw std::runtime_error if the scalar is less than the
+   *    specified lower bound
+   */
+  template <typename Ret, typename LB, typename... Sizes,
+            require_not_std_vector_t<Ret>* = nullptr>
+  auto read_lb(const LB& lb, Sizes... sizes) {
+    auto ret = this->read<Ret>(sizes...);
+    using stan::math::check_greater_or_equal;
+    using stan::math::value_of;
+    check_greater_or_equal("io deserializer", "Lower Bound", value_of(ret),
+                           value_of(lb));
+    return ret;
+  }
```

Comment:
Oh I see things like this exist in the old reader: https://github.com/stan-dev/stan/blob/develop/src/stan/io/reader.hpp#L602

Do we need these any more? I don't think they're used as-in.

Reply:
I looked quickly at a stanc2 generated .hpp file and I don't think these are used there as well.

Reply:
So right now the stanc3 compiler doesn't use these but it will when it uses this function. Right now it just calls the constrain check directly as a loop, but by doing it here we can use the vectorized stuff

Reply:
So is this like transform inits? Or write array or something? I'm just not sure where the constrained parameters exist as a flattened vector.

Reply:
(Like, the constrained parameters exist as a flattened vector, to be read out into multiple variables)

Reply:
It may be be easier to point to the old code that we're replacing than explaining it with words. I made a gist [here](https://gist.github.com/bbbales2/0053578d33035ab25560dd5e2b5a6fb1) with generated code. Which part needs the read + check constrains? It is for this model:

```stan
data {
  vector<lower = 0.0>[3] x_d;
}

transformed data {
  vector<lower = 0.0>[3] x_tf = x_d;
}

parameters {
  vector<lower = 0.0>[3] x_p;
}

transformed parameters {
  vector<lower = 0.0>[3] x_tp = x_p;
}

generated quantities {
  vector<lower = 0.0>[3] x_gq = x_p;
}
```

Reply:
@SteveBronder you hear a voice from your computer

"Steve", it says, "answer meeeee"

It is your old trusty friend, stan-dev/stan #3013, discussion r588771274

Reply:
lol apologies, yeah actually I fleshed your example out a bit more and I think your right we don't need these anymore?

```stan
// deserializer not used
/*
data {
  real<lower=0> x_scal;
  vector<lower = 0.0>[3] x_v;
  matrix<lower = 0.0>[3, 3] x_m;
  matrix<lower = 0.0>[3, 3] x_vm[3];
}

transformed data {
  real<lower=0> x_tscal = x_scal;
  vector<lower = 0.0>[3] x_tv = x_v;
  matrix<lower = 0.0>[3, 3] x_tm = x_m;
  matrix<lower = 0.0>[3, 3] x_tvm[3] = x_vm;
}
*/
parameters {
  real<lower=0> x_param_scal;
  vector<lower = 0.0>[3] x_param_v;
  matrix<lower = 0.0>[3, 3] x_param_m;
  matrix<lower = 0.0>[3, 3] x_param_vm[3];
  cholesky_factor_corr[3] x_param_corr;
  cholesky_factor_corr[3] x_param_vcorr[3];
  matrix<offset = x_param_scal>[3, 3] x_opvm[3];
}

transformed parameters {
  real<lower=0> x_t_param_scal = x_param_scal;
  vector<lower = 0.0>[3] x_t_param_v = x_param_v;
  matrix<lower = 0.0>[3, 3] x_t_param_m = x_param_m;
  matrix<lower = 0.0>[3, 3] x_t_param_vm[3] = x_param_vm;
  cholesky_factor_corr[3] x_t_param_corr = x_param_corr;
  cholesky_factor_corr[3] x_t_param_vcorr[3] = x_param_vcorr;
  matrix<offset = x_param_scal>[3, 3] x_topvm[3];
}

generated quantities {
  vector<lower = 0.0>[3] x_gq = x_t_param_v;
}
```

Looking at the [gist](https://gist.github.com/SteveBronder/3e326e14db5b93586a40f0ba37b17d32) yeah I think your 100% right that we can get rid of these.

Tagging @rybern just to sanity check, is there anywhere that we currently call `reader` (which I think is just `log_prob` and `write_array`) where we need to check that an object satisfies a given constrain but doesn't perform the transform? idt so and that let's me cut out a ton of code here

Reply:
Also @rybern just wanted to sanity check, the deserializer thing here just takes code like

```cpp
      std::vector<Eigen::Matrix<local_scalar_t__, -1, -1>> x_opvm;
      x_opvm = std::vector<Eigen::Matrix<local_scalar_t__, -1, -1>>(
          3, Eigen::Matrix<local_scalar_t__, -1, -1>(3, 3));
      stan::math::fill(x_opvm, DUMMY_VAR__);

      current_statement__ = 7;
      for (int sym1__ = 1; sym1__ <= 3; ++sym1__) {
        current_statement__ = 7;
        assign(x_opvm, cons_list(index_uni(sym1__), nil_index_list()),
               in__.matrix(3, 3), "assigning variable x_opvm");
      }
      current_statement__ = 7;
      for (int sym1__ = 1; sym1__ <= 3; ++sym1__) {
        current_statement__ = 7;
        for (int sym2__ = 1; sym2__ <= 3; ++sym2__) {
          current_statement__ = 7;
          for (int sym3__ = 1; sym3__ <= 3; ++sym3__) {
            current_statement__ = 7;
            if (jacobian__) {
              current_statement__ = 7;
              assign(
                  x_opvm,
                  cons_list(index_uni(sym1__),
                            cons_list(index_uni(sym2__),
                                      cons_list(index_uni(sym3__),
                                                nil_index_list()))),
                  stan::math::offset_multiplier_constrain(
                      rvalue(x_opvm,
                             cons_list(index_uni(sym1__),
                                       cons_list(index_uni(sym2__),
                                                 cons_list(index_uni(sym3__),
                                                           nil_index_list()))),
                             "x_opvm"),
                      x_param_scal, 1, lp__),
                  "assigning variable x_opvm");
            } else {
              current_statement__ = 7;
              assign(
                  x_opvm,
                  cons_list(index_uni(sym1__),
                            cons_list(index_uni(sym2__),
                                      cons_list(index_uni(sym3__),
                                                nil_index_list()))),
                  stan::math::offset_multiplier_constrain(
                      rvalue(x_opvm,
                             cons_list(index_uni(sym1__),
                                       cons_list(index_uni(sym2__),
                                                 cons_list(index_uni(sym3__),
                                                           nil_index_list()))),
                             "x_opvm"),
                      x_param_scal, 1),
                  "assigning variable x_opvm");
            }
          }
        }
      }
```

and let's the compiler generate the following instead

```cpp
// just using an alias for readability
using std_vec_mat = std::vector<Eigen::Matrix<double, -1, -1>>;
std_vec_mat x_opvm = in__.read_offset_multiplier<std_vec_mat, Jacobian>(x_param_scal, 1, lp__, 3, 3, 3);
```

Is that cool for what you are looking to do as well? For tuples you would call

```cpp
using a_tuple_t = std::tuple<local_scalar_t__, std_vec_mat>;
 a_tuple_t foo = std::make_tuple(in__.read_lb<local_scalar_t, Jacobian>(lb_val, lp),
   in__.read_offset_multiplier<std_vec_mat, Jacobian>(offset, mult, lp, 3, 3, 3))
```

Reply:
Yup should work!

Reply:
@SteveBronder I actually missed your first question - 95% sure there's nowhere we do reads and not constraints

Reply:
If it turns out we made a mistake cutting the checks then we can add them back in another pull without too much trouble. Code exists here in the history weeee